### PR TITLE
[WIP] Additions for "subquery" ephemeral type

### PR DESCRIPTION
**Pull request by jamesrguy**
Code for issue [1248](https://github.com/fishtown-analytics/dbt/issues/1248)

Let me know if this should be a different branch, etc..

### CODE REVIEW

**Comment by drewbanin** [comment]
Thanks for the PR @jamesrguy! This looks great! I played around with this locally, and I think there are a couple of tweaks that could be made here.

I imagine that the use case is to make _all_ ephemeral models materialized with subqueries, and one good way to do that is in the `dbt_project.yml` file. Eg:
```
models:
  ephemeral_type: subquery
```

^ This isn't currently supported by the code in your PR, but it's a pretty easy fix to make. Just add a named config [here](https://github.com/fishtown-analytics/dbt/blob/b17d70679fcecae0872ea22b71b92fa6613c17a0/core/dbt/utils.py#L40) and then specify it as a "clobber" field [here](https://github.com/fishtown-analytics/dbt/blob/dev/stephen-girard/core/dbt/parser/source_config.py#L23).

If you make these changes, then users will be able to define all of their ephemeral models to be rendered as subquries with one config change :)

Will also drop a comment inline in the code. Happy to discuss! Thanks again for the PR :)

File: core/dbt/context/runtime.py [hunk 1]
```diff
@@ -47,13 +47,16 @@ def do_ref(*args):
...
         if is_ephemeral:
-            model.set_cte(target_model_id, None)
-            return adapter.Relation.create(
-                type=adapter.Relation.CTE,
-                identifier=add_ephemeral_model_prefix(
-                    target_model_name)).quote(identifier=False)
+            is_subquery = (get_ephemeral_type(target_model) == 'subquery')
+            if is_subquery:
+                return '(\n{}\n)'.format(target_model.compiled_sql)
+            else:
```
Comment (drewbanin) [inline core/dbt/context/runtime.py]: can we do:
```
ephemeral_type = get_ephemeral_type(target_model)
if ephemeral_type == 'subquery':
..
elif ephemeral_type == 'cte' or ephemeral_type is None:
...
else:
  # raise an error here for an invalid value
```

You can raise an error with:
```
dbt.exceptions.raise_compiler_error(msg, target_model)
```

And the message can be something like:
> Invalid value for 'ephemeral_type'. Got '<value>' but expected on of 'cte' or 'subquery'

**Comment by jamesrguy** [comment]
There is another issue with this. In my model code where I ref() a ephemeral subquery model, i'm writing the subquery alias in that code, as is necessary to be able to refer to it in the select part of the query. This however leaves the alias blank in the DBT tests. So I think something extra is needed here to abstract the alias. I'll have to think about this a bit more if nobody has any immediate ideas.

**Comment by drewbanin** [comment]
@jamesrguy really good point. I'm unsure what the best way to proceed is, but you're right, this is something we'll have to figure out to get this merged

**Comment by drewbanin** [comment]
I'm closing this as stale, but happy to revisit if anyone is interested in picking it up!

**Pull request closed without merging**